### ISSUE TITLE
CREDENTIAL: Use of FederatedCredential for federation protocols is unclear

### ISSUE THREAD

**Issue by adrianhopebailie** [w3c/webappsec#266]
Some federated identities are delivered via a standardised protocol like OpenID Connect as opposed to a documented protocol for a specific identity provider. It is not necessary for the exact IdP to be known. The RP only needs to know that they conform to the standard.

It's unclear how an app that supports such a federated identity would request it via `credentials.get()`?

My best guess at how this may work is:

```javascript
navigator.credentials.get({
  "types": [ "PasswordCredential", "FederatedCredential" ],
  "filter": {
      "federations": [ "https://www.facebook.com", "https://openid.net",  ]
  }
}).then(
    function(credential) {
      if (!credential)
        return;

      if (credential instanceof FederatedCredential) {
        switch (credential.federation) {
        case "https://facebook.com/":
          // Use Facebook’s SDK, a la https://developers.facebook.com/docs/facebook-login/login-flow-for-web/#logindialog
          FB.login(function (response) {
            if (response.status === "authorized") {
              // Can now use FB.api() calls
            } else {
              // Explain to the user that we would really like them to click "Log me in"
            }
          });
          break;
                
        case "https://openid.net":
          // Use a generic OpenID Client library and the credential.id
          OpenId.login(credential.id, function (response) {
            // In this case the client needs to follow some discovery protocol 
            //  to figure out the IdP from the identifier
          });
          break;

        // ...
        }
      } else {
        credential.send("https://example.com/loginEndpoint")
          .then(function (response) { ... })
          .catch(function (response) { ... });
      }
});
```

However this breaks the rules around federation identifiers in section [3.3.1. Identifying Federations](https://w3c.github.io/webappsec/specs/credentialmanagement/#identifying-federations)

**Comment by mikewest** [comment]
Let me rephrase, just so I know we're on the same page:

Some relying parties don't care what origin they speak to in order to authenticate a user. All they care about is the language used to speak to that provider. So, we need a mechanism by which an RP can accept "any IdP that speaks OpenID", and by which an IdP can assert "I speak OpenID"?

**Comment by adrianhopebailie** [comment]
@mikewest Nicely put, yes

**Comment by adrianhopebailie** [comment]
I had some thoughts on this:

* There are two different properties implied by the `federation` property:
  1. The federated identity provider
  1. The protocol used for federation
* In many cases (those handled by the spec today) the protocol is implied by the provider so this is fine for those but this is not always the case.
* The protocol could be indicated by using a specialisation of the `FederatedCredential` that is defined by a version of the federation protocol standard that is aware of this API (i.e. OpenID could define that when using OpenIDConnect with this API then a `OpenIdConnectCredential` must be defined with `type = "openidconnect"`). The problem with this is that the `get()` method becomes unwieldy. How do i, as a relying party, execute `get()` and say that I support OpenID Connect but also generic `FederationCredential` objects with `federation = "https://facebook.com"`. In it's current form the query will exclude all `OpenIdConnectCredental`s because they don't have `federation = "https://facebook.com"`.

*Proposal*

1. Change the `federation` property to be called `providerOrigin` or `provider`. The term "identity provider" is well understood.
1. Change the filtering to something along the lines of my [previous suggestion](https://github.com/w3c/webappsec/issues/274#issuecomment-94061897) that allows per-type filter parameters

**Comment by mikewest** [comment]
1. Changing `federation` to `provider` sounds fine.

2. As we've established that the SDK-requiring federations have a protocol implied by the federation itself, I think we can simply add `protocol` to the `FederatedCredential` directly, and wait to extend out to something like a `OpenIDCredential` (which, presumably, would have methods to trigger some user agent behavior to mint tokens). That's more "Future Work" than v1, in my opinion.

**Comment by mikewest** [comment]
WDYT of the two attached patches?

**Comment by adrianhopebailie** [comment]
@mikewest LGTM :+1:

**Comment by mikewest** [comment]
Ok, then let's close this out, and deal with future work in other issues. :)
